tests/serialization/test_parquet.py

```diff
@@ -459,3 +461,55 @@ def test_reset_dict_cols(store):
     only_a_reset = _reset_dictionary_columns(table, exclude=["colB"]).schema
     assert not pa.types.is_dictionary(only_a_reset.field("col").type)
     assert pa.types.is_dictionary(only_a_reset.field("colB").type)
+
+
+def test_retry_on_IOError(monkeypatch, caplog, store):
+    """
+    See https://github.com/JDASoftwareGroup/kartothek/issues/407 :
+    We are testing a retry-workaround for the above issue here. Once the issue is resolved,
+    this test and the workaround can be removed.
+    """
+
+    def patched__restore_dataframe(**kwargs):
```

Comment:
This test shows that a `IOError` will be rewritten to a `ParquetReadError` and that we'll get three logs, not that we actually retry.

```python
count = 0
def patched__restore_dataframe(**kwargs):
    nonlocal count
    count += 1
    if count > 1:
        return
    else:
        raise IOError
```

Reply:
Technically true, I thought that asserting for `len(caplog.records) == MAX_NB_RETRIES`  **and** the corresponding log message would be a strong enough condition for the current setup...

Reply:
I just want to ensure this is not an "all or nothing"

Reply:
Done.